tms/models/tms_order.py

```diff
@@ -0,0 +1,336 @@
+
+    def _default_stage_id(self):
+        stage = self.env["tms.stage"].search(
+            [
+                ("stage_type", "=", "order"),
+            ],
+            order="sequence asc",
+            limit=1,
+        )
+        if stage:
+            return stage
+        raise ValidationError(_("You must create an TMS order stage first."))
+
+    @api.depends("route")
+    def _compute_route_id(self):
+        for order in self:
+            if order.route:
+                order.origin_id = None
+                order.destination_id = None
+            else:
+                order.route_id = None
+
+    @api.depends("scheduled_duration", "scheduled_date_start")
+    def _compute_scheduled_date_end(self):
+        for record in self:
+            if record.scheduled_date_start:
+                record.scheduled_date_end = record.scheduled_date_start + timedelta(
+                    hours=record.scheduled_duration
+                )
+
+    @api.depends("scheduled_date_end", "route_id")
+    def _compute_scheduled_duration(self):
+        for record in self:
+            if (
+                record.scheduled_date_end and record.scheduled_date_start
+            ) and not record.route_id:
+                difference = record.scheduled_date_end - record.scheduled_date_start
+                record.scheduled_duration = difference.total_seconds() / 3600
+            elif record.route_id:
+                if (
+                    record.route_id.estimated_time_uom.id
+                    == self.env.ref("uom.product_uom_day").id
+                ):
+                    record.scheduled_duration = record.route_id.estimated_time * 24
+                else:
+                    record.scheduled_duration = record.route_id.estimated_time
+            else:
+                record.scheduled_duration = 0.0
+
+    @api.depends("duration", "date_start")
+    def _compute_date_end(self):
+        for record in self:
+            if record.date_start:
+                record.date_end = record.date_start + timedelta(hours=record.duration)
+
+    @api.depends("date_end")
+    def _compute_duration(self):
+        for record in self:
+            if record.date_end and record.date_start:
+                difference = record.date_end - record.date_start
+                record.duration = difference.total_seconds() / 3600
+
+    @api.depends("tms_team_id.driver_ids", "crew_id.driver_ids")
+    def _compute_driver_ids_domain(self):
+        all_drivers = self.env["tms.driver"].search([])
+        all_driver_ids = all_drivers.ids
+        for order in self:
+            order.driver_ids_domain = [(6, 0, all_driver_ids)]
+            if order.tms_team_id:
+                order.driver_ids_domain = [(6, 0, order.tms_team_id.driver_ids.ids)]
+            if order.crew_id:
+                order.driver_ids_domain = [(6, 0, order.crew_id.driver_ids.ids)]
+
+    driver_ids_domain = fields.Many2many(
```

Comment:
FYI, you can also use `fields.Binary`to represent a real domain like `[("id", "in", all_vehicles.ids)]`

Reply:
See: https://github.com/odoo/odoo/blob/17.0/addons/account/models/account_tax.py#L1738